client/session.go

```diff
@@ -744,27 +806,29 @@ func (s *session) WriteTagged(
 	unit xtime.Unit,
 	annotation []byte,
 ) error {
-	return errNotImplemented
+	w := s.writeAttemptPool.Get()
+	w.args.attemptType = taggedWriteAttemptType
+	w.args.namespace, w.args.id, w.args.tags = namespace, id, tags
+	w.args.t, w.args.value, w.args.unit, w.args.annotation =
+		t, value, unit, annotation
+	err := s.writeRetrier.Attempt(w.attemptFn)
+	s.writeAttemptPool.Put(w)
+	return err
 }
 
 func (s *session) writeAttempt(
+	wType writeAttemptType,
 	namespace, id ident.ID,
+	inputTags ident.TagIterator,
 	t time.Time,
 	value float64,
 	unit xtime.Unit,
 	annotation []byte,
 ) error {
-	var (
-		enqueued int32
-		majority = atomic.LoadInt32(&s.majority)
-		// NB(prateek): We retain an individual copy of the namespace, ID per
-		// writeState, as each writeState tracks the lifecycle of it's resources in
-		// use in the various queues. Tracking per writeAttempt isn't sufficient as
-		// we may enqueue multiple writeStates concurrently depending on retries
-		// and consistency level checks.
-		nsID = s.idPool.Clone(namespace)
-		tsID = s.idPool.Clone(id)
-	)
+	if wType != untaggedWriteAttemptType && wType != taggedWriteAttemptType {
+		// should never happen
+		return errUnknownWriteAttemptType
+	}
```

Comment:
Since this is already checked in `writeAttemptWithRLock(...)` perhaps it's not necessary?  (the switch on the `wtype`.

Reply:
Yeah this was more paranoia than necessary. Removed in the next patch.